lib/posix-process/signal/sigtimedwait.c

```diff
@@ -0,0 +1,57 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <signal.h>
+#include <stdlib.h>
+#include <sys/time.h>
+
+#include <uk/essentials.h>
+#include <uk/syscall.h>
+#include <uk/timeutil.h>
+
+#include "process.h"
+#include "signal.h"
+
+UK_SYSCALL_R_DEFINE(int, rt_sigtimedwait,
+		    const sigset_t *__unused, set,
+		    siginfo_t *, info,
+		    const struct timespec *__unused, timeout,
+		    size_t __unused, sigsetsize)
+{
+	struct posix_thread *pthread;
+	struct uk_signal *sig;
+
+	if (unlikely(!set))
+		return -EINVAL;
+
+	if (unlikely(sigsetsize != sizeof(sigset_t)))
+		return -EINVAL;
+
+	pthread = tid2pthread(uk_gettid());
+
+	/* TODO protect concurrent access by deliver_pending_proc() */
+	uk_sigcopyset(&pthread->signal->sigwait_set, set);
+
+	/* If a signal in the set is already pending return immediately */
+	if ((sig = pprocess_signal_next_pending_t(pthread)))
+		goto out;
+
+	if (timeout)
```

Comment:
As per POSIX:
```
If the timespec structure pointed to by timeout is zero-valued and if none of the signals
specified by set are pending, then sigtimedwait( ) shall return  immediately with an error.
If timeout is the null pointer, the behavior is unspecified.
```

So I assume that the above case with `pprocess_signal_next_pending_t` should also check for `timeout` value and return an error?

And in the case `timeout` is just `NULL` we should throw an assertion?

Reply:
I follow the behavior of linux. From sigwaitinfo(2):
> POSIX  leaves the meaning of a NULL value for the timeout argument of sigtimedwait() unspecified,
> permitting the possibility that this has the same meaning as a call to sigwaitinfo(), and indeed this is
> what is done on Linux.